frontend/controller/actions/identity.js

```diff
@@ -369,24 +360,128 @@ export default (sbp('sbp/selectors/register', {
+
+          // Decrypt the state encryption key
+          stateEncryptionKeyS = decrypt(stateKeyEncryptionKey, encryptedStateEncryptionKey, stateEncryptionKeyId)
+
+          // Compute the key ID of the decrypted key and verify that it holds
+          // the expected value
+          const stateEncryptionKeyIdActual = keyId(stateEncryptionKeyS)
+          if (stateEncryptionKeyIdActual !== stateEncryptionKeyId) {
+            throw new Error(`Invalid state key ID: expected ${stateEncryptionKeyId} but got ${stateEncryptionKeyIdActual}`)
+          }
+        }
+
+        // Now, attempt to decrypt the state
+        const state = JSON.parse(decrypt(stateEncryptionKeyS, data, identityContractID || ''))
+
+        // If decryption succeeded, set the necessary values for the login state
+        // and the session storage
+        extraLoginAttributes.stateEncryptionKeyId = stateEncryptionKeyId
+        extraLoginAttributes.salt = salt
+        extraLoginAttributes.encryptedStateEncryptionKey = encryptedStateEncryptionKey
+
+        // Saving the state encryption key in session storage is necessary
+        // for functionality such as refreshing the page to work
+        await sbp('gi.db/settings/save', stateEncryptionKeyId, stateEncryptionKeyS)
+
+        return state
+      }).catch(async (e) => {
+        // If we haven't got a password, we can't proceed deriving keys or
+        // encrypting, so we give up at this point
+        if (!password) { throw e }
+
+        // If the state was empty, this isn't necessarily an error
+        if (!(e instanceof EmptyState)) {
+          console.warn('Error while retrieving local state', e)
+        }
+
+        // Create the necessary keys
+        // First, we generate the state encryption key
+        const stateEncryptionKey = keygen(CURVE25519XSALSA20POLY1305)
+        const stateEncryptionKeyId = keyId(stateEncryptionKey)
+        const stateEncryptionKeyS = serializeKey(stateEncryptionKey, true)
+
+        // Once we have the state encryption key, we generate a salt
+        const salt = generateSalt()
+
+        // We use the salt, the state encryption key ID and the password to
+        // derive a key to encrypt the state encryption key
+        // This key is not stored anywhere, but is used for reconstructing
+        // the state on a fresh session
+        const stateKeyEncryptionKey = await deriveKeyFromPassword(CURVE25519XSALSA20POLY1305, password, salt + stateEncryptionKeyId)
+
+        // Once everything is place, set the attributes for extraLoginAttributes
+        extraLoginAttributes.stateEncryptionKeyId = stateEncryptionKeyId
+        extraLoginAttributes.salt = salt
+        extraLoginAttributes.encryptedStateEncryptionKey = encrypt(stateKeyEncryptionKey, stateEncryptionKeyS, stateEncryptionKeyId)
+
+        // Save the state encryption key to local storage
+        await sbp('gi.db/settings/save', stateEncryptionKeyId, stateEncryptionKeyS)
+      })
```

Comment:
Please keep the login function as simple as it was before by moving all of this code elsewhere.

Reply:
Can be put in a new selector called `'gi.db/settings/loadEncrypted'` (and create a `'gi.db/settings/saveEncrypted'` below)